consensus/voteset/binary_voteset.go

```diff
@@ -0,0 +1,149 @@
+package voteset
+
+import (
+	"github.com/pactus-project/pactus/crypto"
+	"github.com/pactus-project/pactus/types/validator"
+	"github.com/pactus-project/pactus/types/vote"
+	"github.com/pactus-project/pactus/util/errors"
+)
+
+type roundVotes struct {
+	// Each vote can have one of 3 possible values: {0,1,Abstain}.
+	voteBoxes  [3]*voteBox
+	allVotes   map[crypto.Address]*vote.Vote
+	votedPower int64
+}
+
+func newRoundVotes() *roundVotes {
+	voteBoxes := [3]*voteBox{}
+	voteBoxes[vote.CPValueZero] = newVoteBox()
+	voteBoxes[vote.CPValueOne] = newVoteBox()
+	voteBoxes[vote.CPValueAbstain] = newVoteBox()
+
+	return &roundVotes{
+		voteBoxes:  voteBoxes,
+		allVotes:   make(map[crypto.Address]*vote.Vote),
+		votedPower: 0,
+	}
+}
+
+func (bs *roundVotes) addVote(v *vote.Vote, power int64) {
+	vb := bs.voteBoxes[v.CPValue()]
+	vb.addVote(v, power)
+	bs.votedPower += power
+}
+
+type BinaryVoteSet struct {
+	*voteSet
+	roundVotes []*roundVotes
+}
+
+func NewCPPreVoteVoteSet(round int16, totalPower int64,
+	validators map[crypto.Address]*validator.Validator,
+) *BinaryVoteSet {
+	voteSet := newVoteSet(vote.VoteTypeCPPreVote, round, totalPower, validators)
+	return newBinaryVoteSet(voteSet)
+}
+
+func NewCPMainVoteVoteSet(round int16, totalPower int64,
+	validators map[crypto.Address]*validator.Validator,
+) *BinaryVoteSet {
+	voteSet := newVoteSet(vote.VoteTypeCPMainVote, round, totalPower, validators)
+	return newBinaryVoteSet(voteSet)
+}
+
+func newBinaryVoteSet(voteSet *voteSet) *BinaryVoteSet {
+	return &BinaryVoteSet{
+		voteSet:    voteSet,
+		roundVotes: make([]*roundVotes, 0, 1),
+	}
+}
+
+func (vs *BinaryVoteSet) mustGetRoundVotes(cpRound int16) *roundVotes {
+	for i := len(vs.roundVotes); i <= int(cpRound); i++ {
+		rv := newRoundVotes()
+		vs.roundVotes = append(vs.roundVotes, rv)
+	}
+
+	return vs.roundVotes[cpRound]
+}
+
+// AllVotes returns a list of all votes in the VoteSet.
+func (vs *BinaryVoteSet) AllVotes() []*vote.Vote {
+	votes := make([]*vote.Vote, 0)
+	for _, rv := range vs.roundVotes {
+		for _, v := range rv.allVotes {
+			votes = append(votes, v)
+		}
+	}
+	return votes
+}
+
+// AddVote attempts to add a vote to the VoteSet. Returns an error if the vote is invalid.
+func (vs *BinaryVoteSet) AddVote(v *vote.Vote) (bool, error) {
+	power, err := vs.voteSet.verifyVote(v)
+	if err != nil {
+		return false, err
+	}
+
+	roundVotes := vs.mustGetRoundVotes(v.CPRound())
+	existingVote, ok := roundVotes.allVotes[v.Signer()]
+	if ok {
+		if existingVote.Hash() != v.Hash() {
+			err = errors.Error(errors.ErrDuplicateVote)
```

Comment:
I think we should return err at line 93

Reply:
We set error and return it later.
Mote that an adversary can double vote. n this case the second vote is bad vote, but we should add it.